File: FSUIKit/Classes/FSReuseTableView.m
//
//  FSReuseTableView.m
//  myhome
//
//  Created by FudonFuchina on 2018/4/5.
//  Copyright © 2018年 fuhope. All rights reserved.
//

#import "FSReuseTableView.h"
#import "MJRefresh.h"

@interface FSReuseTableView ()<UITableViewDelegate,UITableViewDataSource>

@end

@implementation FSReuseTableView

- (instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self bestCellDesignViews];
    }
    return self;
}

- (void)bestCellDesignViews{
    if (!_tableView) {
        _tableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, self.bounds.size.width, self.bounds.size.height) style:UITableViewStyleGrouped];
        _tableView.delegate = self;
        _tableView.dataSource = self;
        _tableView.estimatedSectionHeaderHeight = 0;
        _tableView.estimatedSectionFooterHeight = 0;
        [self addSubview:_tableView];
        
    }
}

- (void)setRefresh_header:(void (^)(void))refresh_header{
    _refresh_header = refresh_header;
    
    if (refresh_header) {
        __weak typeof(self)this = self;
        _tableView.mj_header= [MJRefreshNormalHeader headerWithRefreshingBlock:^{
            if (this.refresh_header) {
                this.refresh_header();
            }
        }];
    }
}

- (void)setRefresh_footer:(void (^)(void))refresh_footer{
    _refresh_footer = refresh_footer;
    
    if (refresh_footer) {
        __weak typeof(self)this = self;
        _tableView.mj_footer = [MJRefreshBackNormalFooter footerWithRefreshingBlock:^{
            if (this.refresh_footer) {
                this.refresh_footer();
            }
        }];
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    if (self.numberOfSections) {
        NSInteger count = self.numberOfSections(tableView);
        return count;
    }
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if (self.numberOfRowsInSection) {
        NSInteger count = self.numberOfRowsInSection(tableView, section);
        return count;
    }
    return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (self.cellForRowAtIndexPath) {
        return self.cellForRowAtIndexPath(tableView, indexPath);
    }
    return nil;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath{
    if (self.willDisplayCell) {
        self.willDisplayCell(cell,indexPath);
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (self.heightForRowAtIndexPath) {
        return self.heightForRowAtIndexPath(tableView,indexPath);
    }
    return 44;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    if (self.heightForHeaderInSection) {
        return self.heightForHeaderInSection(tableView,section);
    }
    return 0;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    if (self.heightForFooterInSection) {
        return self.heightForFooterInSection(tableView, section);
    }
    return 0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if (self.didSelectRowAtIndexPath) {
        self.didSelectRowAtIndexPath(tableView, indexPath);
    }
}

- (void)endRefresh{
    [_tableView.mj_header endRefreshing];
    [_tableView.mj_footer endRefreshing];
}

- (void)deleteRefresh{
    [_tableView.mj_header removeFromSuperview];
    [_tableView.mj_footer removeFromSuperview];
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: FSUIKit/Classes/Swift/FSReuseTableView.swift
// FSReuseTableViewS.swift
// Translated from FSReuseTableView.h/m
// Note: MJRefresh dependency removed - use native UIRefreshControl instead

import UIKit

@objcMembers
public class FSReuseTableView: UIView {
    
    public var tableView: UITableView!
    
    public var refreshHeader: (() -> Void)? {
        didSet {
            if refreshHeader != nil {
                let refreshControl = UIRefreshControl()
                refreshControl.addTarget(self, action: #selector(handleRefreshHeader), for: .valueChanged)
                tableView.refreshControl = refreshControl
            }
        }
    }
    
    public var refreshFooter: (() -> Void)?
    
    public var numberOfSections: ((UITableView) -> Int)?
    public var numberOfRowsInSection: ((UITableView, Int) -> Int)?
    public var cellForRowAtIndexPath: ((UITableView, IndexPath) -> UITableViewCell?)?
    public var willDisplayCell: ((UITableViewCell, IndexPath) -> Void)?
    public var heightForRowAtIndexPath: ((UITableView, IndexPath) -> CGFloat)?
    public var heightForHeaderInSection: ((UITableView, Int) -> CGFloat)?
    public var heightForFooterInSection: ((UITableView, Int) -> CGFloat)?
    public var didSelectRowAtIndexPath: ((UITableView, IndexPath) -> Void)?
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        bestCellDesignViews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        bestCellDesignViews()
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        tableView.frame = bounds
    }
    
    private func bestCellDesignViews() {
        tableView = UITableView(frame: CGRect(x: 0, y: 0, width: bounds.width, height: bounds.height), style: .grouped)
        tableView.delegate = self
        tableView.dataSource = self
        tableView.estimatedSectionHeaderHeight = 0
        tableView.estimatedSectionFooterHeight = 0
        addSubview(tableView)
    }
    
    @objc private func handleRefreshHeader() {
        refreshHeader?()
    }
    
    public func endRefresh() {
        tableView.refreshControl?.endRefreshing()
    }
    
    public func deleteRefresh() {
        tableView.refreshControl = nil
    }
}

extension FSReuseTableViewS: UITableViewDelegate, UITableViewDataSource {
    public func numberOfSections(in tableView: UITableView) -> Int {
        return numberOfSections?(tableView) ?? 0
    }
    
    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return numberOfRowsInSection?(tableView, section) ?? 0
    }
    
    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return cellForRowAtIndexPath?(tableView, indexPath) ?? UITableViewCell()
    }
    
    public func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        willDisplayCell?(cell, indexPath)
    }
    
    public func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return heightForRowAtIndexPath?(tableView, indexPath) ?? 44
    }
    
    public func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return heightForHeaderInSection?(tableView, section) ?? 0
    }
    
    public func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return heightForFooterInSection?(tableView, section) ?? 0
    }
    
    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        didSelectRowAtIndexPath?(tableView, indexPath)
    }
}
